### PR TITLE
feat(notifications): Add Reads to NotificationSettingsManager

### PR DESCRIPTION
**Pull request by mgaeta**
This PR adds new functions for reading values from the new NotificationSetting table. Until now the NotificationsSettings manager only implemented methods to write and delete values.

These new functions make it easier to query for lists of NotificationSettings:
- `get_settings`
- `get_for_user_by_projects`
- `get_for_users_by_parent`

These functions interpret the DB to get a list of users that have opted into getting notifications:
- `filter_to_subscribed_users`
- `get_notification_recipients`

And these functions manipulate lists of NotificationsSettings to answers questions about user preferences:
- `should_be_participating`
- `should_user_be_notified`
- `transform_to_notification_settings_by_parent_id`
- `transform_to_notification_settings_by_user`

### CODE REVIEW

File: src/sentry/notifications/helpers.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+def transform_to_notification_settings_by_user(
+    notification_settings: Iterable,
+    users: Iterable,
+) -> Mapping[any, Mapping[NotificationScopeType, NotificationSettingOptionValues]]:
+    """
+    Given a unorganized list of notification settings, create a mapping of
+    users to a map of notification scopes to setting values.
+    """
+    actor_mapping = {user.actor: user for user in users}
```
Comment (wedamija) [inline src/sentry/notifications/helpers.py:76]: Same comment re `attach_foreignkey`/`actor_id`

File: src/sentry/notifications/helpers.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    notification_settings_by_user = defaultdict(dict)
+    for notification_setting in notification_settings:
+        user = actor_mapping.get(notification_setting.target)
```
Comment (wedamija) [inline src/sentry/notifications/helpers.py:79]: Is it valid/possible for user to be `None` here?

File: src/sentry/notifications/helpers.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
+from collections import defaultdict
```
Comment (wedamija) [inline src/sentry/notifications/helpers.py:1]: This file looks like it's basically entirely typed - could you add it to `mypy.ini` (assuming it's not throwing a bunch of type errors for other reasons)

File: src/sentry/notifications/manager.py [hunk 4]
```diff
@@ -71,74 +33,57 @@ def get_settings(
...
-                scope_type=scope_type,
-                scope_identifier=scope_identifier,
-                target=target,
-            ).first()
-            or NotificationSettingOptionValues.DEFAULT
+        from sentry.models import UserOption
+
+        setting_option = self.find_settings(
+            provider, type, user, team, project, organization
+        ).first()
```
Comment (wedamija) [inline src/sentry/notifications/manager.py:51]: Careful about using `.first()`. If your queryset doesn't have a sort order specified, it adds one in, which can break index use. If all you want is the first setting, it can be safer to use `list(qs)[0]`

File: src/sentry/api/serializers/models/user_notifications.py [hunk 5]
```diff
@@ -1,35 +1,48 @@
...
+from sentry.models.integration import ExternalProviders
+from sentry.models import NotificationSetting, UserOption
+from sentry.notifications.legacy_mappings import (
+    get_type_from_fine_tuning_key,
+    map_notification_settings_to_legacy,
+)
+from sentry.notifications.types import FineTuningAPIKey, NotificationScopeType
+
+
+def handle_legacy(notification_type: FineTuningAPIKey, users: Iterable) -> Iterable:
```
[the remark this answers is not in the document]
Comment (mgaeta) [reply src/sentry/api/serializers/models/user_notifications.py:14]: I've been running into circular dependency problems because the functions are top-level in the file. What would be really cool is if the `User` model inherited a `User` or `UserType` placeholder type.

File: src/sentry/api/serializers/models/user_notifications.py [hunk 6]
```diff
@@ -1,35 +1,48 @@
...
+    """ For EMAIL and REPORTS, check UserOptions. """
+    filter_args = {}
+    if notification_type == FineTuningAPIKey.EMAIL:
```
[the remark this answers is not in the document]
Comment (mgaeta) [reply src/sentry/api/serializers/models/user_notifications.py:17]: There are five options for `FineTuningAPIKey`. Three will read from `NotificationSetting` and these two still need to read from `UserOption`.
```
class FineTuningAPIKey(Enum):
    ALERTS = "alerts" 		# NotificationSetting
    DEPLOY = "deploy"		# NotificationSetting
    EMAIL = "email"			# UserOption
    REPORTS = "reports"		# UserOption
    WORKFLOW = "workflow" 	# NotificationSetting
```

File: src/sentry/api/endpoints/user_notification_details.py [hunk 7]
```diff
@@ -24,11 +25,17 @@ def get_attrs(self, item_list, user, *args, **kwargs):
             ).select_related("user")
         )
 
-        results = defaultdict(list)
+        actor_mapping = {user.actor: user for user in item_list}
```
[the remark this answers is not in the document]
Comment (mgaeta) [reply src/sentry/api/endpoints/user_notification_details.py:28]: This is a great point, I'm going to change everything to use IDs.

File: src/sentry/notifications/helpers.py (hunk 2 above)
Comment (mgaeta) [reply src/sentry/notifications/helpers.py]: We looked up the NotificationSettings by this `actor_id` and because we got the ID from the list of users, it should always be in that list. 
I made it paranoid to handle a missing User to appease the type checker.

File: src/sentry/notifications/manager.py (hunk 4 above)
Comment (mgaeta) [reply src/sentry/notifications/manager.py:51]: The `unique_together` constraint should make it impossible to get more than one value, but I don't want to accidentally make the query slower. I'll update it and add a comment.

File: src/sentry/models/organization.py [hunk 8]
```diff
@@ -164,10 +164,16 @@ def save(self, *args, **kwargs):
...
-    def delete(self):
+    def delete(self, **kwargs):
+        from sentry.models import NotificationSetting
+
         if self.is_default:
             raise Exception("You cannot delete the the default organization.")
-        return super().delete()
+
+        # There is no foreign key relationship so we have to manually cascade.
+        NotificationSetting.objects.remove_for_organization(self)
```
Comment (ceorourke) [inline src/sentry/models/organization.py:174]: This and some other files seem to have duplicated code changes that are in https://github.com/getsentry/sentry/pull/24360/

File: src/sentry/notifications/manager.py [hunk 9]
```diff
@@ -72,74 +33,57 @@ def get_settings(
...
-        value = (  # NOQA
-            self.filter(
-                provider=provider.value,
-                type=type.value,
-                scope_type=scope_type,
-                scope_identifier=scope_identifier,
-                target=target,
-            ).first()
-            or NotificationSettingOptionValues.DEFAULT
+        from sentry.models.useroption import UserOption
```
Comment (ceorourke) [inline src/sentry/notifications/manager.py:47]: why was this moved down here? import issues?

**Review by ceorourke — APPROVED** [review]

File: src/sentry/notifications/manager.py (hunk 9 above)
Comment (mgaeta) [reply src/sentry/notifications/manager.py:47]: Yeah trying to prevent problems with circular imports.
